ansible/roles/metrics/files/process-cloudflare/process-cloudflare.js

```diff
@@ -0,0 +1,172 @@
+      return 'ppc64le'
+    } else if (fileType.indexOf('ppc64') >= 0) {
+      return 'ppc64'
+    } else if (fileType.indexOf('s390x') >= 0) {
+      return 's390x'
+    }
+  }
+
+  if (os === 'win') {
+    // we get here for older .msi files and node.exe files
+    if (winArch && winArch.indexOf('x64') >= 0) {
+      // could be 'x64' or 'win-x64'
+      return 'x64'
+    } else {
+      // could be 'win-x86' or ''
+      return 'x86'
+    }
+  }
+
+  return ''
+}
+
+const logTransformStream = new Transform({
+  writableObjectMode: true,
+  transform (chunk, encoding, callback) {
+    if (chunk.ClientRequestMethod !== 'GET' ||
+        chunk.EdgeResponseStatus < 200 ||
+        chunk.EdgeResponseStatus >= 300) {
+      return callback()
+    }
+
+    if (chunk.EdgeResponseBytes < 1024) { // unreasonably small for something we want to measure
+      return callback()
+    }
+
+    if (!extensionRe.test(chunk.ClientRequestPath)) { // not a file we care about
+      return callback()
+    }
+
+    const requestPath = chunk.ClientRequestPath.replace(/\/\/+/g, '/')
+    const uriMatch = requestPath.match(uriRe)
+    if (!uriMatch) { // what is this then?
+      return callback()
+    }
+
+    const path = uriMatch[1]
+    const pathVersion = uriMatch[4]
+    const file = uriMatch[5]
+    const winArch = uriMatch[6]
+    const fileVersion = uriMatch[8]
+    const fileType = uriMatch[9]
+
+    let version = ''
+    // version can come from the filename or the path, filename is best
+    // but it may not be there (e.g. node.exe) so fall back to path version
+    if (versionRe.test(fileVersion)) {
+      version = fileVersion
+    } else if (versionRe.test(pathVersion)) {
+      version = pathVersion
+    }
+
+    const os = determineOS(path, file, fileType)
+    const arch = determineArch(fileType, winArch, os)
+
+    const line = []
+    line.push(strftime('%Y-%m-%d', new Date(chunk.EdgeStartTimestamp / 1000 / 1000))) // date
+    line.push(chunk.ClientCountry.toUpperCase()) // country
+    line.push('') // state/province, derived from chunk.EdgeColoCode probably
+    line.push(chunk.ClientRequestPath) // URI
+    line.push(version) // version
+    line.push(os) // os
+    line.push(arch) // arch
+    line.push(chunk.EdgeResponseBytes)
+
+    this.push(`${line.join(',')}\n`)
+
+    callback()
+  }
+})
+
+
+exports.processLogs = (data, context, callback) => {
+  console.log('Node version is: ' + process.version);
+  const file = data;
+  bucketName = file.bucket;
+  fileName = file.name;
+  console.log("DATA " + data);
+  console.log("BUCKET " + bucketName);
+  console.log("FILENAME " + fileName);
+  processedFile = fileName.split(".")[0];
+  processedFile = processedFile.split("_")[0].concat("_", processedFile.split("_")[1]); 
+  console.log("PROCESSEDFILENAME " + processedFile);
+
+  storage.bucket(bucketName).file(file.name).createReadStream()
+  .on('error', function(err) { console.error(err) })
```

Comment:
Each one of the pipes below here could emit an error you won't see. Copy this line and paste it below each pipe and you may catch an error that's otherwise being lost. The alternative is to use util.pipeline() which handles error propagation nicely.

Reply:
@rvagg After doing this the log now reads like this 
<img width="733" alt="image" src="https://user-images.githubusercontent.com/26796102/89199382-5eccf800-d5a6-11ea-9132-9e47ca4d4c36.png">

Reply:
Did you try removing the `e` from the jsonStream error callback? Perhaps we're getting a line somewhere that we can't parse properly for some reason but should be able to ignore - might be worth trying. It could be that an error in either jsonStream or logTransformStream is closing the downstream streams and split2 can't push anymore in.

`util.pipeline()` is really worth trying here, not just for the error handling but also for the nice cleanup it does and managing these complicated events.

Reply:
Yep removed the `e` changed the script to this
```
exports.processLogs = (data, context, callback) => {
  console.log('Node version is: ' + process.version);
  const file = data;
  bucketName = file.bucket;
  fileName = file.name;
  console.log("DATA " + data);
  console.log("BUCKET " + bucketName);
  console.log("FILENAME " + fileName);
  processedFile = fileName.split(".")[0];
  processedFile = processedFile.split("_")[0].concat("_", processedFile.split("_")[1]); 
  console.log("PROCESSEDFILENAME " + processedFile);

  pipeline(
    storage.bucket(bucketName).file(file.name).createReadStream(),
    split2(),
    jsonStream,
    logTransformStream,
    storage.bucket('processed-logs-nodejs').file(processedFile).createWriteStream({resumable: false},
    (err) => {
      if (err) {
        console.log("PIPELINE HAS FAILED", err)
        
      } else {
        console.log("PIPELINE SUCCESS")
        callback()
      }
    }
  )
  )
```


been running for halfhour with no errors yet but more interestingly it constantly timesout - no matter how big I set the timeout, So I assume im not calling the `callback()` in the right place so the function doesn't end

Reply:
Feel free to keep on pushing your changes to this branch, at least we can have a discussion around the same code in that case.

One of the tricky things about custom streams is that you have to always be aware of calling the callback on `transform()` and `flush()` on _any path_ through your code. Failure to do so will cause weird errors that involve back-pressure that has a delayed effect so it can be hard to track down. So the `jsonStream` and `logTransformStream` are worth studying to find codepaths that may not properly trigger a `callback()`, possibly even on the case of an uncaught exception which will just halt that execution path.

You do have one obvious instance of a failure to call callback though, at the bottom in the failure case, maybe it should be `callback(err)` after the `console.log("PIPELINE HAS FAILED", err)`, just for completeness. But presumably you'd see that somewhere in the output (not necessarily _right_ at the end though).

Reply:
ive pushed my code up, ive realised that I dont see any pipeline finish message - either pass or fail but the files _seem_ to be roughly correct in size (its not an exact science) so I assume the processing is working but the pipeline is just not ending untill GCP kills it.

Reply:
rebased as @BethGriggs pointed out I had a bracket in the wrong place - time outs now stop but we are back to the weird behaviour of before with empty out files. the pipeline is reporting a success however @rvagg.